clang/test/SemaSYCL/CudaCompat/delayed-diags.cu

```diff
@@ -0,0 +1,43 @@
+// RUN: %clang_cc1 %s -fsycl-is-host -fsycl-cuda-compatibility \
+// RUN:   -internal-isystem %S/../../SemaCUDA/Inputs \
+// RUN:   -internal-isystem %S/../Inputs \
+// RUN:   -fsycl-targets=nvptx64-nvidia-cuda -triple x86_64-unknown-linux \
+// RUN:   -o %t -verify
+// RUN: %clang_cc1 %s -fsycl-is-device -fsycl-cuda-compatibility \
+// RUN:   -internal-isystem %S/../../SemaCUDA/Inputs \
+// RUN:   -internal-isystem %S/../Inputs \
+// RUN:   -fsycl-targets=nvptx64-nvidia-cuda -triple x86_64-unknown-linux\
+// RUN:   -o %t -verify
+
+// Check that delayed diagnostics are done
+// according to the SYCL logic and not the CUDA one.
+// fsycl-cuda-compatibility doesn't enable the CUDA device mode,
+// this leaves the diagnostic logic for CUDA it is processing the host side
+// and triggers SYCL's delayed diagnostics.
+
+// expected-no-diagnostics
+
+#include "cuda.h"
+#include "sycl.hpp"
+
+struct Foo {
+  Foo();
+};
+
+Foo& get() {
+  // Checks no error is raised for non const static variable.
+  static Foo f;
+  return f;
+}
```

Comment:
Why would a diagnostic be required for a non-const static local variable? The function isn't explicitly annotated as host, device, or both (I thus expect it to be implicitly host-only), nor is it used within the TU.

Reply:
Right, but why would there ever be an expectation of a diagnostic here? A non-const static local variable is not an error in any configuration that I'm aware of.

Reply:
> A non-const static local variable is not an error in any configuration that I'm aware of.

It is in device code (see last bullet point in https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:language.restrictions.kernels). During the processing of the file, this decl triggers a delayed diagnostic and this is asserting it is never actually raised as `get` is never emitted during the device compilation.

Reply:
Thank you, I wasn't aware of that requirement (or had forgotten it). This makes perfect sense now.